metadata-ingestion/tests/integration/azure_ad/test_azure_ad.py

```diff
@@ -308,6 +315,121 @@ def load_test_resources(test_resources_dir):
     )
 
 
+def get_current_checkpoint_from_pipeline(
+    pipeline: Pipeline,
+) -> Optional[Checkpoint[GenericCheckpointState]]:
+    azure_ad_source = cast(AzureADSource, pipeline.source)
+    return azure_ad_source.get_current_checkpoint(
+        azure_ad_source.stale_entity_removal_handler.job_id
+    )
+
+
+def stateful_ingest(
```

Comment:
first, this should be named test_stateful_ingest

second, this looks like a lot of copy-pasted code. could you refactor it to reduce the copying?

Reply:
done